src/bus/webSocketConnection.js

```diff
@@ -173,32 +146,196 @@ class WebSocketConnection {
+    return this._registerSingleMessageHandler('Publish', {
+      service_id: serviceClientId,
+      channel,
+      message
+    });
+  }
+
+  /**
+   * Subscribes to a specific channel on the message bus and handles messages as they are received. When the handler is
+   * called, the message is automatically acknowledged after the message completes except whenever an Error is thrown.
+   * The user can also programmatically control when the message is acknowledged by calling `ack` at any time.
+   *
+   * @param {string} serviceClientId Client ID of the message bus service
+   * @param {string} channel Message bus channel the message is being sent to
+   * @param {string} [group] A unique identifier for the subscriber that can be shared between connections
+   * @param {function} handler A function that gets invoked with every received message
+   * @param {function} errorHandler A function that gets invoked with every error
+   *
+   * @example
+   *   contxtSdk.bus.connect('4f0e51c6-728b-4892-9863-6d002e61204d')
+   *     .then((webSocket) => {
+   *       webSocket.subscribe('GCXd2bwE9fgvqxygrx2J7TkDJ3ef', 'feed:1', 'test-sub', (message) => {
+   *         console.log('Message received: ', message);
+   *       }, (error) => {
+   *         console.log('Error received: ', error);
+   *       });
+   *     });
+   *
+   * @example
+   *   contxtSdk.bus.connect('4f0e51c6-728b-4892-9863-6d002e61204d')
+   *     .then((webSocket) => {
+   *       webSocket.subscribe('GCXd2bwE9fgvqxygrx2J7TkDJ3ef', 'feed:1', 'test-sub', (message, ack) => {
+   *         console.log('Message received: ', message);
+   *
+   *         ack();
+   *       }, (error) => {
+   *         console.log('Error received: ', error);
+   *       });
+   *     });
+   *
+   * @example
+   *   contxtSdk.bus.connect('4f0e51c6-728b-4892-9863-6d002e61204d')
+   *     .then((webSocket) => {
+   *       webSocket.subscribe('GCXd2bwE9fgvqxygrx2J7TkDJ3ef', 'feed:1', (message) => {
+   *         return db.save(message);
+   *       }, (error) => {
+   *         console.log('Error received: ', error);
+   *       });
+   *     });
+   *
+   * @example
+   *   contxtSdk.bus.connect('4f0e51c6-728b-4892-9863-6d002e61204d')
+   *     .then((webSocket) => {
+   *       webSocket.subscribe('GCXd2bwE9fgvqxygrx2J7TkDJ3ef', 'feed:1', (message, ack) => {
+   *         return db.save(message)
+   *           .then(ack)
+   *           .then(() => {
+   *             // additional processing
+   *           });
+   *       }, (error) => {
+   *         console.log('Error received: ', error);
+   *       });
+   *     });
+   */
+  subscribe(serviceClientId, channel, group, handler, errorHandler) {
+    if (!serviceClientId) {
+      return Promise.reject(
+        new Error('A service client id is required for subscribing')
+      );
+    }
+
+    if (!channel) {
+      return Promise.reject(new Error('A channel is required for subscribing'));
+    }
+
+    if (typeof group === 'function') {
```

Comment:
Just from an organizational standpoint, what do you think about moving this down below all the other `error` checking? Just so you can read through all the error cases quickly and then when all those pass, you get to the actionable stuff

Reply:
You're right! Moving it below would affect how `handler` and `errorHandler` are defined, but I can move it to the top. The code is to allow `group` to be a completely optional parameter, but not at the end of the function because an API like `subscribe(serviceId, channel, (message) => {}, (error) => {}, group)` looks a little odd to me. Great suggestion, should help the readability a bit.

Reply:
Ah, yes. That's right 👍